### PR TITLE
feature: make main contract pausable

### PR DESCRIPTION
**Pull request by zoruka**

### CODE REVIEW

File: contracts/contracts/FleekPausable.sol
Comment (EmperorOrokuSaki) [inline contracts/contracts/FleekPausable.sol:16]: Maybe we can rename this to something more descriptive like `PauseStatusChanged`

File: contracts/contracts/FleekPausable.sol
Comment (EmperorOrokuSaki) [inline contracts/contracts/FleekPausable.sol:21]: Also for this maybe we can use `PausableStatusChanged`

File: contracts/contracts/FleekPausable.sol
Comment (EmperorOrokuSaki) [inline contracts/contracts/FleekPausable.sol:24]: We need to think about when the contract is not pausable first. For example, when a transaction is initiated before the pause, it will revert if the contract's status changes.
At the same time, we need to consider the possibility of bad actors finding a balance of enough gas to keep the transaction pending but not enough priority gas for it to be picked up faster. They can make the contract impossible to pause externally and that's a bad position to be in.

File: contracts/contracts/FleekPausable.sol
Comment (EmperorOrokuSaki) [inline contracts/contracts/FleekPausable.sol:61]: maybe rename to `isPaused`

File: contracts/contracts/FleekPausable.sol
Comment (EmperorOrokuSaki) [inline contracts/contracts/FleekPausable.sol:68]: maybe rename to `isPausable`

File: contracts/contracts/FleekPausable.sol
Comment (EmperorOrokuSaki) [inline contracts/contracts/FleekPausable.sol:97]: Wondering what's the reason for the `virtual` keyword here and some of the other functions. These functions being overridable is not a good decision.

File: contracts/contracts/FleekPausable.sol
Comment (EmperorOrokuSaki) [inline contracts/contracts/FleekPausable.sol:30]: Doesn't this line contradict the comment on line 27?

File: contracts/contracts/FleekERC721.sol
Comment (EmperorOrokuSaki) [inline contracts/contracts/FleekERC721.sol:539]: the `whenNotPaused` modifier is forgotten here.
Also, we need to discuss when we will use this feature and decide on what functions we want to work on when the contract is paused.

File: contracts/contracts/FleekERC721.sol
Comment (EmperorOrokuSaki) [inline contracts/contracts/FleekERC721.sol:604]: Is these section going to be affected by the acl refactor?

**Review by EmperorOrokuSaki — CHANGES_REQUESTED** [review]
I have some questions about this feature and its implementation.

I will review the tests when I understand the feature better based on the questions I have asked here in my review.

File: contracts/contracts/FleekERC721.sol
Comment (zoruka) [reply contracts/contracts/FleekERC721.sol:604]: Yes. I will need to fix #114 after this got merged.

File: contracts/contracts/FleekERC721.sol
Comment (zoruka) [reply contracts/contracts/FleekERC721.sol:539]: It is on `_beforeTokenTransfer` function that is called for mint, transfer and burn functions.

File: contracts/contracts/FleekPausable.sol
Comment (zoruka) [reply contracts/contracts/FleekPausable.sol:30]: Yeah, going to fix the comment

File: contracts/contracts/FleekPausable.sol
Comment (zoruka) [reply contracts/contracts/FleekPausable.sol:97]: That's correct, it was added by the copilot and I forgot to remove

File: contracts/contracts/FleekPausable.sol
Comment (zoruka) [reply contracts/contracts/FleekPausable.sol:61]: for both function names I have followed the pattern of https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/security/Pausable.sol, but I agree to change it

File: contracts/contracts/FleekPausable.sol
Comment (zoruka) [reply contracts/contracts/FleekPausable.sol:24]: Agreed, this needs to be discussed. cc @jsonsivar

Comment (jsonsivar) [reply contracts/contracts/FleekPausable.sol:24]: I'm not sure I understand the issue here yet. The _canPause is like a feature flag that only Fleek controller can call (and later can be set by a governance contract or multisig). The only time we will need to use this pause feature imo is if it's an emergency and we need to pause it to protect peoples assets. In that example like a vulnerability is discovered with our contracts or EVM e.g., I don't think it matters of any transactions initiated before the pause transaction fails right? I could be missing something here.

File: contracts/contracts/FleekPausable.sol
Comment (jsonsivar) [reply contracts/contracts/FleekPausable.sol:30]: Wait I think it should start in unpaused state right? Not paused but can pause would be the init state imo.

File: contracts/contracts/FleekPausable.sol
Comment (jsonsivar) [inline contracts/contracts/FleekPausable.sol:97]: Should we do this like a toggle where we can switch it back too? Or create a separate function that sets canPause to false. This way its a feature flag that could be set back and forth as governance decides (and at the moment Fleek controller).

**Review by jsonsivar — CHANGES_REQUESTED** [review]

File: contracts/contracts/FleekERC721.sol
Comment (zoruka) [reply contracts/contracts/FleekERC721.sol:536]: the modifier added to the function `_beforeTokenTransfer` that is called for mint, transfer and burn function.
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/ERC721.sol#L302
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/ERC721.sol#L268
https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC721/ERC721.sol#L337

File: contracts/contracts/FleekPausable.sol
Comment (zoruka) [reply contracts/contracts/FleekPausable.sol:97]: I think I did not get the question. The function should always invert the state? I just think this way is safer because you cannot set the flag to a state you don't want by accident.

File: contracts/contracts/FleekPausable.sol
Comment (jsonsivar) [reply contracts/contracts/FleekPausable.sol:30]: But in that case I think at the start we would just not deploy the contract until release moment.  I can't think of a scenario where we would want to release it but pause it at start so nobody can use it.  I noticed the Zeppelin example starts unpaused too.  Also if we leave it paused we need a second transaction just to start it.  Maybe we can make the constructor take the start pause state so we can be flexible with it?

File: contracts/contracts/FleekPausable.sol
Comment (jsonsivar) [reply contracts/contracts/FleekPausable.sol:97]: But we should still have a way to turn of off after. Right now once you make it pausable you can never go back?

File: contracts/contracts/FleekERC721.sol
Comment (jsonsivar) [reply contracts/contracts/FleekERC721.sol:536]: Ah gotcha. Does that make the call from here redundant then https://github.com/fleekxyz/non-fungible-apps/pull/110/files#diff-3fc5d3120a8e235b37eb66c89e4759fcead60f2b3ea2a4e5fb664892ce541b0cR183? Cuz mint also calls `_beforeTokenTransfer` right?

File: contracts/contracts/FleekPausable.sol
Comment (zoruka) [reply contracts/contracts/FleekPausable.sol:30]: I think we can have it unpaused by default, I don't think we need a flexible start for it tbh. I'm going to change it

File: contracts/contracts/FleekPausable.sol
Comment (zoruka) [reply contracts/contracts/FleekPausable.sol:97]: No, we can change it whenever we want. It just checks if you are trying to set the same value as it is, in this case it reverts.

File: contracts/test/foundry/FleekERC721/AccessControl.t.sol
Comment (jsonsivar) [inline contracts/test/foundry/FleekERC721/AccessControl.t.sol:428]: Should this go into another test file? Just wondering if access control is the best place or might be hard to tell that from the name. Maybe the pausable file or main erc721 test file.

**Review by jsonsivar — APPROVED** [review]
LGTM

File: contracts/test/foundry/FleekERC721/AccessControl.t.sol
Comment (zoruka) [reply contracts/test/foundry/FleekERC721/AccessControl.t.sol:428]: Here it is just testing if the address is able or not to call the function and not the function run itself, the same as other functions in this test suit.

File: contracts/contracts/FleekERC721.sol
Comment (zoruka) [reply contracts/contracts/FleekERC721.sol:536]: I think the way it is displayed by github is miss leading you to see it wrong, the modifier is not in `mint` function

**Comment by jsonsivar** [comment]
Looks like merging develop into this broke CI.  I think from the error details the ABI is out of sync with the subgraph.yaml.

File: contracts/contracts/FleekERC721.sol
Comment (jsonsivar) [reply contracts/contracts/FleekERC721.sol:522]: You're right, I'll mark this as resolved.

**Comment by jsonsivar** [comment]
This looks good now. Just needs approval from @EmperorOrokuSaki and it should be good to go.

**Pull request merged**
